src/Compilers/CSharp/Test/Syntax/Parsing/LambdaParameterParsingTests.cs

```diff
@@ -580,5 +581,135 @@ public void HangingLambdaParsing_Bug14167()
             EOF();
         }
 
+        [Fact]
+        public void TestLambdaWithNullValidation()
+        {
+            UsingDeclaration("Func<string, string> func1 = x! => x + \"1\";");
```

Comment:
formatter needs to ensure it never collapses `!` and `=` in any circumstances.

need tests for `x!, y! =>`